## Data cleaning assignment

In [1]:
import pandas as pd 
import numpy as np
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm','Budapest_PaRis', 'Brussels_londOn'],
                   'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
                   'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )','12. Air France', '"Swiss Air"']})

In [6]:
df.head()

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

1. Some values in the the FlightNumber column are missing. These numbers are
meant to increase by 10 with each row so 10055 and 10075 need to be put in
place. Fill in these missing numbers and make the column an integer column
(instead of a float column).

In [3]:
# Making a copy of dataframe for further operations.
df_a= df

In [4]:
x,y = df_a.FlightNumber.loc[[0]].values[0]+10, df_a.FlightNumber.loc[[4]].values[0]-10

df_a.FlightNumber.loc[[1]]=df_a.FlightNumber.loc[[1]].fillna(x)

df_a.FlightNumber.loc[[3]]=df_a.FlightNumber.loc[[3]].fillna(y)

# convert flot type to int type
df_a.FlightNumber = df_a.FlightNumber.astype(int)

D:\Anaconda\envs\test_env\lib\site-packages\pandas\core\indexing.py:205: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


In [5]:
df_a.head()

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

2. The From_To column would be better as two separate columns! Split each
string on the underscore delimiter _ to give a new temporary DataFrame with
the correct values. Assign the correct column names to this temporary
DataFrame.

In [7]:
df_temp= df_a.From_To.str.split('_', expand=True)
df_temp.columns = ['From', 'To']
print("Output")
df_temp.head()

Output


From         To
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn

3. Notice how the capitalisation of the city names is all mixed up in this
temporary DataFrame. Standardise the strings so that only the first letter is
uppercase (e.g. "londON" should become "London".)

In [9]:
df_temp.From = df_temp.From.str.title()
df_temp.To = df_temp.To.str.title()
print("Output")
df_temp.head()

Output


From         To
0    London      Paris
1    Madrid      Milan
2    London  Stockholm
3  Budapest      Paris
4  Brussels     London

4. Delete the From_To column from df and attach the temporary DataFrame
from the previous questions.

In [11]:
df_a=df_a.drop("From_To", axis=1).join(df_temp)
print("Output")
df_a.head()

Output


FlightNumber  RecentDelays              Airline      From         To
0         10045      [23, 47]               KLM(!)    London      Paris
1         10055            []    <Air France> (12)    Madrid      Milan
2         10065  [24, 43, 87]  (British Airways. )    London  Stockholm
3         10075          [13]       12. Air France  Budapest      Paris
4         10085      [67, 32]          "Swiss Air"  Brussels     London

5. In the RecentDelays column, the values have been entered into the
DataFrame as a list. We would like each first value in its own column, each

second value in its own column, and so on. If there isn't an Nth value, the value
should be NaN.
Expand the Series of lists into a DataFrame named delays, rename the columns
delay_1, delay_2, etc. and replace the unwanted RecentDelays column in df
with delays.

In [12]:
df_delays=pd.DataFrame(df_a["RecentDelays"].values.tolist())
df_delays.columns = ['delay_{}'.format(n) for n in range(1, len(df_delays.columns)+1)]
df_delays

delay_1  delay_2  delay_3
0     23.0     47.0      NaN
1      NaN      NaN      NaN
2     24.0     43.0     87.0
3     13.0      NaN      NaN
4     67.0     32.0      NaN

In [13]:
df_a=df_a.drop("RecentDelays" , axis=1).join(df_delays)
print("Output")
df_a

Output


FlightNumber              Airline      From         To  delay_1  delay_2  \
0         10045               KLM(!)    London      Paris     23.0     47.0   
1         10055    <Air France> (12)    Madrid      Milan      NaN      NaN   
2         10065  (British Airways. )    London  Stockholm     24.0     43.0   
3         10075       12. Air France  Budapest      Paris     13.0      NaN   
4         10085          "Swiss Air"  Brussels     London     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN